pkg/maps/metricsmap/metricsmap.go

```diff
@@ -137,54 +149,133 @@ func (vs Values) Bytes() uint64 {
 	return b
 }
 
-func updateMetric(getCounter func() (prometheus.Counter, error), newValue float64) {
-	counter, err := getCounter()
+// metricsMapCollector implements Prometheus Collector interface
+type metricsmapCollector struct {
+	droppedCountDesc *prometheus.Desc
+	droppedByteDesc  *prometheus.Desc
+	forwardCountDesc *prometheus.Desc
+	forwardByteDesc  *prometheus.Desc
+
+	// eBPF code seems to expose multiple reasons for forwarded metrics
+	// as opposed to what is stated in bpf/lib/metrics.h comments.
+	// IterateWithCallback iterates through BPF map for each reason and direction.
+	// Since we do not have "reason" label on forwarded metrics, we would end up collecting
+	// same forwarded metric multiple times which is not allowed by prometheus client.
+	// See https://github.com/prometheus/client_golang/issues/242
+	//
+	// forwarded and dropped metrics maps are used to sum all values by desired set of labels
+	forwardedMetricsMap map[forwardLabels]metricValues
+	droppedMetricsMap   map[dropLabels]metricValues
+}
+
+type forwardLabels struct {
+	direction string
+}
+
+type dropLabels struct {
+	direction string
+	reason    string
+}
+
+type metricValues struct {
+	bytes float64
+	count float64
+}
+
+func newMetricsMapCollector() prometheus.Collector {
+	return &metricsmapCollector{
+		droppedMetricsMap:   make(map[dropLabels]metricValues),
+		forwardedMetricsMap: make(map[forwardLabels]metricValues),
+		droppedByteDesc: prometheus.NewDesc(
+			prometheus.BuildFQName(metrics.Namespace, metrics.SubsystemDatapath, "drop_bytes_total"),
+			"Total dropped bytes, tagged by drop reason and ingress/egress direction",
+			[]string{metrics.LabelDropReason, metrics.LabelDirection}, nil,
+		),
+		droppedCountDesc: prometheus.NewDesc(
+			prometheus.BuildFQName(metrics.Namespace, metrics.SubsystemDatapath, "drop_count_total"),
+			"Total dropped packets, tagged by drop reason and ingress/egress direction",
+			[]string{metrics.LabelDropReason, metrics.LabelDirection}, nil,
+		),
+		forwardCountDesc: prometheus.NewDesc(
+			prometheus.BuildFQName(metrics.Namespace, metrics.SubsystemDatapath, "forward_count_total"),
+			"Total forwarded packets, tagged by ingress/egress direction",
+			[]string{metrics.LabelDirection}, nil,
+		),
+		forwardByteDesc: prometheus.NewDesc(
+			prometheus.BuildFQName(metrics.Namespace, metrics.SubsystemDatapath, "forward_bytes_total"),
+			"Total forwarded bytes, tagged by ingress/egress direction",
+			[]string{metrics.LabelDirection}, nil,
+		),
+	}
+}
+
+func (mc *metricsmapCollector) Collect(ch chan<- prometheus.Metric) {
+	err := Metrics.IterateWithCallback(func(key *Key, values *Values) {
+		if key.IsDrop() {
```

Comment:
We could perhaps dry this up a bit by creating a metricsmap generic type and hang methods to upsert or even exporting metrics on the map vs using functions? i.e
```go
type metricsMap[k labels] map[k]metricValues

func (m metricsMap[labels]) upsert(l labels, b, c float64) {
	m[l] = m[l].inc(b, c)
}
// Then...
err := Metrics.IterateWithCallback(func(key *Key, values *Values) {
		if key.IsDrop() {
		  m.droppedMetricMap.upsert(dropLabels{...}, values.Bytes, value.Counts)
		} else {
		  m.forwardedMetricMap.upsert(forwardLabels{...}, values.Bytes, value.Counts)
		}
                ...
}
```

Reply:
@derailed Thanks for review! 
I've created a generic `promMetrics` map with upsert() method as suggested. Code in a callback function looks bit cleaner now. 
Regarding exporting, I agree that it would be better if we could have `m.droppedMetricMap.export` and `m.forwardedMetricMap.export` instead those two for loops at the end of callback function, but in that case we would have to associate prometheus.Desc with labels and values. Since we have two values under same key (labels) in metrics map I don't see any elegant way to do this without rewriting most of the logic, if you have any hints, I'm willing to try :)

Reply:
@carnerito I think what you have now is indeed fine.